File: sdk/src/utils/type-chart.ts
import { MonsterType } from "../types";

/**
 * Full 17-type effectiveness chart.
 * typeChart[attacker][defender] = multiplier
 * 2.0 = super effective, 0.5 = not very effective, 0.0 = immune, 1.0 = neutral
 */
const typeChart: number[][] = buildTypeChart();

function buildTypeChart(): number[][] {
  const T = MonsterType;
  const chart: number[][] = Array.from({ length: 17 }, () =>
    new Array(17).fill(1.0)
  );

  function se(atk: MonsterType, def: MonsterType): void {
    chart[atk][def] = 2.0;
  }
  function nve(atk: MonsterType, def: MonsterType): void {
    chart[atk][def] = 0.5;
  }
  function imm(atk: MonsterType, def: MonsterType): void {
    chart[atk][def] = 0.0;
  }

  // Normal
  nve(T.Normal, T.Rock);
  nve(T.Normal, T.Steel);
  imm(T.Normal, T.Ghost);

  // Fire
  se(T.Fire, T.Grass);
  se(T.Fire, T.Ice);
  se(T.Fire, T.Bug);
  se(T.Fire, T.Steel);
  nve(T.Fire, T.Fire);
  nve(T.Fire, T.Water);
  nve(T.Fire, T.Rock);
  nve(T.Fire, T.Dragon);

  // Water
  se(T.Water, T.Fire);
  se(T.Water, T.Ground);
  se(T.Water, T.Rock);
  nve(T.Water, T.Water);
  nve(T.Water, T.Grass);
  nve(T.Water, T.Dragon);

  // Electric
  se(T.Electric, T.Water);
  se(T.Electric, T.Flying);
  nve(T.Electric, T.Electric);
  nve(T.Electric, T.Grass);
  nve(T.Electric, T.Dragon);
  imm(T.Electric, T.Ground);

  // Grass
  se(T.Grass, T.Water);
  se(T.Grass, T.Ground);
  se(T.Grass, T.Rock);
  nve(T.Grass, T.Fire);
  nve(T.Grass, T.Grass);
  nve(T.Grass, T.Poison);
  nve(T.Grass, T.Flying);
  nve(T.Grass, T.Bug);
  nve(T.Grass, T.Dragon);
  nve(T.Grass, T.Steel);

  // Ice
  se(T.Ice, T.Grass);
  se(T.Ice, T.Ground);
  se(T.Ice, T.Flying);
  se(T.Ice, T.Dragon);
  nve(T.Ice, T.Fire);
  nve(T.Ice, T.Water);
  nve(T.Ice, T.Ice);
  nve(T.Ice, T.Steel);

  // Fighting
  se(T.Fighting, T.Normal);
  se(T.Fighting, T.Ice);
  se(T.Fighting, T.Rock);
  se(T.Fighting, T.Dark);
  se(T.Fighting, T.Steel);
  nve(T.Fighting, T.Poison);
  nve(T.Fighting, T.Flying);
  nve(T.Fighting, T.Psychic);
  nve(T.Fighting, T.Bug);
  imm(T.Fighting, T.Ghost);

  // Poison
  se(T.Poison, T.Grass);
  nve(T.Poison, T.Poison);
  nve(T.Poison, T.Ground);
  nve(T.Poison, T.Rock);
  nve(T.Poison, T.Ghost);
  imm(T.Poison, T.Steel);

  // Ground
  se(T.Ground, T.Fire);
  se(T.Ground, T.Electric);
  se(T.Ground, T.Poison);
  se(T.Ground, T.Rock);
  se(T.Ground, T.Steel);
  nve(T.Ground, T.Grass);
  nve(T.Ground, T.Bug);
  imm(T.Ground, T.Flying);

  // Flying
  se(T.Flying, T.Grass);
  se(T.Flying, T.Fighting);
  se(T.Flying, T.Bug);
  nve(T.Flying, T.Electric);
  nve(T.Flying, T.Rock);
  nve(T.Flying, T.Steel);

  // Psychic
  se(T.Psychic, T.Fighting);
  se(T.Psychic, T.Poison);
  nve(T.Psychic, T.Psychic);
  nve(T.Psychic, T.Steel);
  imm(T.Psychic, T.Dark);

  // Bug
  se(T.Bug, T.Grass);
  se(T.Bug, T.Psychic);
  se(T.Bug, T.Dark);
  nve(T.Bug, T.Fire);
  nve(T.Bug, T.Fighting);
  nve(T.Bug, T.Poison);
  nve(T.Bug, T.Flying);
  nve(T.Bug, T.Ghost);
  nve(T.Bug, T.Steel);

  // Rock
  se(T.Rock, T.Fire);
  se(T.Rock, T.Ice);
  se(T.Rock, T.Flying);
  se(T.Rock, T.Bug);
  nve(T.Rock, T.Fighting);
  nve(T.Rock, T.Ground);
  nve(T.Rock, T.Steel);

  // Ghost
  se(T.Ghost, T.Psychic);
  se(T.Ghost, T.Ghost);
  nve(T.Ghost, T.Dark);
  imm(T.Ghost, T.Normal);

  // Dragon
  se(T.Dragon, T.Dragon);
  nve(T.Dragon, T.Steel);
  imm(T.Dragon, T.Normal); // Actually neutral, but we keep fairy-less gen logic
  // Correction: Dragon vs Normal is neutral
  chart[T.Dragon][T.Normal] = 1.0;

  // Dark
  se(T.Dark, T.Psychic);
  se(T.Dark, T.Ghost);
  nve(T.Dark, T.Fighting);
  nve(T.Dark, T.Dark);

  // Steel
  se(T.Steel, T.Ice);
  se(T.Steel, T.Rock);
  nve(T.Steel, T.Fire);
  nve(T.Steel, T.Water);
  nve(T.Steel, T.Electric);
  nve(T.Steel, T.Steel);

  return chart;
}

/**
 * Get the type effectiveness multiplier for a single attacking type vs single defending type.
 */
export function getTypeEffectiveness(
  attackType: MonsterType,
  defendType: MonsterType
): number {
  return typeChart[attackType][defendType];
}

/**
 * Get combined effectiveness for an attack type against a dual-typed defender.
 */
export function getCombinedEffectiveness(
  attackType: MonsterType,
  defendTypes: MonsterType[]
): number {
  let multiplier = 1.0;
  for (const defType of defendTypes) {
    multiplier *= typeChart[attackType][defType];
  }
  return multiplier;
}

/**
 * Check if a type matchup is super effective (>1.0).
 */
export function isSuperEffective(
  attackType: MonsterType,
  defendTypes: MonsterType[]
): boolean {
  return getCombinedEffectiveness(attackType, defendTypes) > 1.0;
}

/**
 * Check if a type matchup is not very effective (<1.0 and >0).
 */
export function isNotVeryEffective(
  attackType: MonsterType,
  defendTypes: MonsterType[]
): boolean {
  const eff = getCombinedEffectiveness(attackType, defendTypes);
  return eff < 1.0 && eff > 0;
}

/**
 * Check if a type matchup is immune (0).
 */
export function isImmune(
  attackType: MonsterType,
  defendTypes: MonsterType[]
): boolean {
  return getCombinedEffectiveness(attackType, defendTypes) === 0;
}

/**
 * Get all types that are super effective against the given types.
 */
export function getSuperEffectiveTypes(
  defendTypes: MonsterType[]
): MonsterType[] {
  const result: MonsterType[] = [];
  for (let i = 0; i < 17; i++) {
    if (isSuperEffective(i as MonsterType, defendTypes)) {
      result.push(i as MonsterType);
    }
  }
  return result;
}

/**
 * Get all types that the given types resist.
 */
export function getResistances(defendTypes: MonsterType[]): MonsterType[] {
  const result: MonsterType[] = [];
  for (let i = 0; i < 17; i++) {
    if (isNotVeryEffective(i as MonsterType, defendTypes)) {
      result.push(i as MonsterType);
    }
  }
  return result;
}

/**
 * Get all types that the given types are immune to.
 */
export function getImmunities(defendTypes: MonsterType[]): MonsterType[] {
  const result: MonsterType[] = [];
  for (let i = 0; i < 17; i++) {
    if (isImmune(i as MonsterType, defendTypes)) {
      result.push(i as MonsterType);
    }
  }
  return result;
}

/**
 * Get the best attacking type against the given defending types.
 */
export function getBestAttackingType(
  defendTypes: MonsterType[]
): { type: MonsterType; effectiveness: number } {
  let bestType = MonsterType.Normal;
  let bestEff = 0;
  for (let i = 0; i < 17; i++) {
    const eff = getCombinedEffectiveness(i as MonsterType, defendTypes);
    if (eff > bestEff) {
      bestEff = eff;
      bestType = i as MonsterType;
    }
  }
  return { type: bestType, effectiveness: bestEff };
}

/**
 * Get a human-readable effectiveness description.
 */
export function getEffectivenessText(multiplier: number): string {
  if (multiplier === 0) return "No effect";
  if (multiplier < 0.5) return "Barely effective";
  if (multiplier === 0.5) return "Not very effective";
  if (multiplier === 1.0) return "Neutral";
  if (multiplier === 2.0) return "Super effective";
  if (multiplier >= 4.0) return "Extremely effective";
  return `${multiplier}x effective`;
}

/**
 * Get the full type chart as a 2D array.
 */
export function getFullTypeChart(): ReadonlyArray<ReadonlyArray<number>> {
  return typeChart;
}

/**
 * Get type name from enum value.
 */
export function getTypeName(type: MonsterType): string {
  return MonsterType[type];
}

File: agents/src/utils/type_chart.py
"""Type effectiveness chart for PXMON - 17 types with full matchup table."""

from enum import Enum
from typing import Dict, Tuple


class MonsterType(Enum):
    NORMAL = "normal"
    FIRE = "fire"
    WATER = "water"
    GRASS = "grass"
    ELECTRIC = "electric"
    ICE = "ice"
    FIGHTING = "fighting"
    POISON = "poison"
    GROUND = "ground"
    FLYING = "flying"
    PSYCHIC = "psychic"
    BUG = "bug"
    ROCK = "rock"
    GHOST = "ghost"
    DRAGON = "dragon"
    DARK = "dark"
    STEEL = "steel"


SUPER_EFFECTIVE = 2.0
NOT_VERY_EFFECTIVE = 0.5
NO_EFFECT = 0.0
NEUTRAL = 1.0

# Attacking type -> Defending type -> multiplier
# Only storing non-neutral matchups to keep it manageable
_CHART: Dict[MonsterType, Dict[MonsterType, float]] = {
    MonsterType.NORMAL: {
        MonsterType.ROCK: NOT_VERY_EFFECTIVE,
        MonsterType.GHOST: NO_EFFECT,
        MonsterType.STEEL: NOT_VERY_EFFECTIVE,
    },
    MonsterType.FIRE: {
        MonsterType.FIRE: NOT_VERY_EFFECTIVE,
        MonsterType.WATER: NOT_VERY_EFFECTIVE,
        MonsterType.GRASS: SUPER_EFFECTIVE,
        MonsterType.ICE: SUPER_EFFECTIVE,
        MonsterType.BUG: SUPER_EFFECTIVE,
        MonsterType.ROCK: NOT_VERY_EFFECTIVE,
        MonsterType.DRAGON: NOT_VERY_EFFECTIVE,
        MonsterType.STEEL: SUPER_EFFECTIVE,
    },
    MonsterType.WATER: {
        MonsterType.FIRE: SUPER_EFFECTIVE,
        MonsterType.WATER: NOT_VERY_EFFECTIVE,
        MonsterType.GRASS: NOT_VERY_EFFECTIVE,
        MonsterType.GROUND: SUPER_EFFECTIVE,
        MonsterType.ROCK: SUPER_EFFECTIVE,
        MonsterType.DRAGON: NOT_VERY_EFFECTIVE,
    },
    MonsterType.GRASS: {
        MonsterType.FIRE: NOT_VERY_EFFECTIVE,
        MonsterType.WATER: SUPER_EFFECTIVE,
        MonsterType.GRASS: NOT_VERY_EFFECTIVE,
        MonsterType.POISON: NOT_VERY_EFFECTIVE,
        MonsterType.GROUND: SUPER_EFFECTIVE,
        MonsterType.FLYING: NOT_VERY_EFFECTIVE,
        MonsterType.BUG: NOT_VERY_EFFECTIVE,
        MonsterType.ROCK: SUPER_EFFECTIVE,
        MonsterType.DRAGON: NOT_VERY_EFFECTIVE,
        MonsterType.STEEL: NOT_VERY_EFFECTIVE,
    },
    MonsterType.ELECTRIC: {
        MonsterType.WATER: SUPER_EFFECTIVE,
        MonsterType.GRASS: NOT_VERY_EFFECTIVE,
        MonsterType.ELECTRIC: NOT_VERY_EFFECTIVE,
        MonsterType.GROUND: NO_EFFECT,
        MonsterType.FLYING: SUPER_EFFECTIVE,
        MonsterType.DRAGON: NOT_VERY_EFFECTIVE,
    },
    MonsterType.ICE: {
        MonsterType.FIRE: NOT_VERY_EFFECTIVE,
        MonsterType.WATER: NOT_VERY_EFFECTIVE,
        MonsterType.GRASS: SUPER_EFFECTIVE,
        MonsterType.ICE: NOT_VERY_EFFECTIVE,
        MonsterType.GROUND: SUPER_EFFECTIVE,
        MonsterType.FLYING: SUPER_EFFECTIVE,
        MonsterType.DRAGON: SUPER_EFFECTIVE,
        MonsterType.STEEL: NOT_VERY_EFFECTIVE,
    },
    MonsterType.FIGHTING: {
        MonsterType.NORMAL: SUPER_EFFECTIVE,
        MonsterType.ICE: SUPER_EFFECTIVE,
        MonsterType.POISON: NOT_VERY_EFFECTIVE,
        MonsterType.FLYING: NOT_VERY_EFFECTIVE,
        MonsterType.PSYCHIC: NOT_VERY_EFFECTIVE,
        MonsterType.BUG: NOT_VERY_EFFECTIVE,
        MonsterType.ROCK: SUPER_EFFECTIVE,
        MonsterType.GHOST: NO_EFFECT,
        MonsterType.DARK: SUPER_EFFECTIVE,
        MonsterType.STEEL: SUPER_EFFECTIVE,
    },
    MonsterType.POISON: {
        MonsterType.GRASS: SUPER_EFFECTIVE,
        MonsterType.POISON: NOT_VERY_EFFECTIVE,
        MonsterType.GROUND: NOT_VERY_EFFECTIVE,
        MonsterType.ROCK: NOT_VERY_EFFECTIVE,
        MonsterType.GHOST: NOT_VERY_EFFECTIVE,
        MonsterType.STEEL: NO_EFFECT,
    },
    MonsterType.GROUND: {
        MonsterType.FIRE: SUPER_EFFECTIVE,
        MonsterType.GRASS: NOT_VERY_EFFECTIVE,
        MonsterType.ELECTRIC: SUPER_EFFECTIVE,
        MonsterType.POISON: SUPER_EFFECTIVE,
        MonsterType.FLYING: NO_EFFECT,
        MonsterType.BUG: NOT_VERY_EFFECTIVE,
        MonsterType.ROCK: SUPER_EFFECTIVE,
        MonsterType.STEEL: SUPER_EFFECTIVE,
    },
    MonsterType.FLYING: {
        MonsterType.GRASS: SUPER_EFFECTIVE,
        MonsterType.ELECTRIC: NOT_VERY_EFFECTIVE,
        MonsterType.FIGHTING: SUPER_EFFECTIVE,
        MonsterType.BUG: SUPER_EFFECTIVE,
        MonsterType.ROCK: NOT_VERY_EFFECTIVE,
        MonsterType.STEEL: NOT_VERY_EFFECTIVE,
    },
    MonsterType.PSYCHIC: {
        MonsterType.FIGHTING: SUPER_EFFECTIVE,
        MonsterType.POISON: SUPER_EFFECTIVE,
        MonsterType.PSYCHIC: NOT_VERY_EFFECTIVE,
        MonsterType.DARK: NO_EFFECT,
        MonsterType.STEEL: NOT_VERY_EFFECTIVE,
    },
    MonsterType.BUG: {
        MonsterType.FIRE: NOT_VERY_EFFECTIVE,
        MonsterType.GRASS: SUPER_EFFECTIVE,
        MonsterType.FIGHTING: NOT_VERY_EFFECTIVE,
        MonsterType.POISON: NOT_VERY_EFFECTIVE,
        MonsterType.FLYING: NOT_VERY_EFFECTIVE,
        MonsterType.PSYCHIC: SUPER_EFFECTIVE,
        MonsterType.GHOST: NOT_VERY_EFFECTIVE,
        MonsterType.DARK: SUPER_EFFECTIVE,
        MonsterType.STEEL: NOT_VERY_EFFECTIVE,
    },
    MonsterType.ROCK: {
        MonsterType.FIRE: SUPER_EFFECTIVE,
        MonsterType.ICE: SUPER_EFFECTIVE,
        MonsterType.FIGHTING: NOT_VERY_EFFECTIVE,
        MonsterType.GROUND: NOT_VERY_EFFECTIVE,
        MonsterType.FLYING: SUPER_EFFECTIVE,
        MonsterType.BUG: SUPER_EFFECTIVE,
        MonsterType.STEEL: NOT_VERY_EFFECTIVE,
    },
    MonsterType.GHOST: {
        MonsterType.NORMAL: NO_EFFECT,
        MonsterType.PSYCHIC: SUPER_EFFECTIVE,
        MonsterType.GHOST: SUPER_EFFECTIVE,
        MonsterType.DARK: NOT_VERY_EFFECTIVE,
    },
    MonsterType.DRAGON: {
        MonsterType.DRAGON: SUPER_EFFECTIVE,
        MonsterType.STEEL: NOT_VERY_EFFECTIVE,
    },
    MonsterType.DARK: {
        MonsterType.FIGHTING: NOT_VERY_EFFECTIVE,
        MonsterType.PSYCHIC: SUPER_EFFECTIVE,
        MonsterType.GHOST: SUPER_EFFECTIVE,
        MonsterType.DARK: NOT_VERY_EFFECTIVE,
    },
    MonsterType.STEEL: {
        MonsterType.FIRE: NOT_VERY_EFFECTIVE,
        MonsterType.WATER: NOT_VERY_EFFECTIVE,
        MonsterType.ELECTRIC: NOT_VERY_EFFECTIVE,
        MonsterType.ICE: SUPER_EFFECTIVE,
        MonsterType.ROCK: SUPER_EFFECTIVE,
        MonsterType.STEEL: NOT_VERY_EFFECTIVE,
    },
}


def get_effectiveness(attack_type: MonsterType, defend_type: MonsterType) -> float:
    """Get the type effectiveness multiplier for an attack."""
    if attack_type not in _CHART:
        return NEUTRAL
    return _CHART[attack_type].get(defend_type, NEUTRAL)


def get_dual_effectiveness(
    attack_type: MonsterType,
    defend_types: Tuple[MonsterType, ...],
) -> float:
    """Get combined effectiveness against a dual-type defender."""
    multiplier = 1.0
    for dtype in defend_types:
        multiplier *= get_effectiveness(attack_type, dtype)
    return multiplier


def get_weaknesses(monster_types: Tuple[MonsterType, ...]) -> list[MonsterType]:
    """Return list of types that are super effective against the given type combo."""
    weaknesses: list[MonsterType] = []
    for atk_type in MonsterType:
        eff = get_dual_effectiveness(atk_type, monster_types)
        if eff >= SUPER_EFFECTIVE:
            weaknesses.append(atk_type)
    return weaknesses


def get_resistances(monster_types: Tuple[MonsterType, ...]) -> list[MonsterType]:
    """Return list of types that are not very effective against the given type combo."""
    resistances: list[MonsterType] = []
    for atk_type in MonsterType:
        eff = get_dual_effectiveness(atk_type, monster_types)
        if 0.0 < eff <= NOT_VERY_EFFECTIVE:
            resistances.append(atk_type)
    return resistances


def get_immunities(monster_types: Tuple[MonsterType, ...]) -> list[MonsterType]:
    """Return list of types that have no effect against the given type combo."""
    immunities: list[MonsterType] = []
    for atk_type in MonsterType:
        eff = get_dual_effectiveness(atk_type, monster_types)
        if eff == NO_EFFECT:
            immunities.append(atk_type)
    return immunities


def best_attack_type(
    available_types: list[MonsterType],
    defender_types: Tuple[MonsterType, ...],
) -> MonsterType:
    """Pick the best attack type from available options against a defender."""
    best_type = available_types[0]
    best_eff = 0.0
    for atype in available_types:
        eff = get_dual_effectiveness(atype, defender_types)
        if eff > best_eff:
            best_eff = eff
            best_type = atype
    return best_type